### PR TITLE
Fix column indices in EnforceDistribution optimizer in Partial AggregateMode

### LINKED ISSUE #4873: panic when GROUP BY column order doesn't match USING column order
**Describe the bug**
[VegaFusion](https://vegafusion.io/) implements the Vega [impute](https://vega.github.io/vega/docs/transforms/impute/) transform with a query that uses a CROSS JOIN to build a table that contains all combinations of the unique values of certain input columns, then a LEFT OUTER JOIN to a subquery that performs a GROUP BY, grouping by these same columns. This LEFT OUTER JOIN includes a USING constraint with the same columns that are used in the GROUP BY subquery. 

When the order of the columns in the GROUP BY doesn't match the order of the columns in the USING constraint, the query panics. If the columns are specified in the same order, the query completes successfully.

**Example**
For example, here is an input table with 3 columns.

```
+------+------+------+
| col1 | col2 | col3 |
+------+------+------+
| A    | 1    | 1    |
| A    | 1    | 1    |
| A    | 2    | 1    |
| A    | 2    | 1    |
| A    | 3    | 1    |
| A    | 3    | 1    |
| A    | 4    | 1    |
| A    | 4    | 1    |
| BB   | 5    | 1    |
| BB   | 5    | 1    |
| BB   | 6    | 1    |
| BB   | 6    | 1    |
+------+------+------+
```
Here is a working query:
```sql
SELECT "col1", "col2", coalesce("sum_col3", 0) as sum_col3
FROM (SELECT DISTINCT "col2" FROM "tbl") AS "subq1"
CROSS JOIN (SELECT DISTINCT "col1" FROM "tbl") AS "subq2"
LEFT OUTER JOIN (SELECT "col1", "col2", sum("col3") as "sum_col3" FROM "tbl" GROUP BY "col1", "col2") AS "subq3"
USING("col1", "col2")
```
```
+------+------+----------+
| col1 | col2 | sum_col3 |
+------+------+----------+
| BB   | 5    | 2        |
| BB   | 1    | 0        |
| BB   | 4    | 0        |
| BB   | 6    | 2        |
| BB   | 3    | 0        |
| A    | 3    | 2        |
| BB   | 2    | 0        |
| A    | 2    | 2        |
| A    | 1    | 2        |
| A    | 5    | 0        |
| A    | 6    | 0        |
| A    | 4    | 2        |
+------+------+----------+
```

The result contains one row for every combination of the unique values in `col1` and `col2`.  The `sum_col3` columns contains the sum of the input `col3` for that combination of `col1` and `col2`, and zeros are filled in for combinations that weren't present in the input table.

Now modify the query by flipping the order of `col1` and `col2` in the USING constraint (From `USING("col1", "col2")` to `USING("col2", "col1")`).  Now the query panics.

```sql
SELECT "col1", "col2", coalesce("sum_col3", 0) as sum_col3
FROM (SELECT DISTINCT "col2" FROM "tbl") AS "subq1"
CROSS JOIN (SELECT DISTINCT "col1" FROM "tbl") AS "subq2"
LEFT OUTER JOIN (SELECT "col1", "col2", sum("col3") as "sum_col3" FROM "tbl" GROUP BY "col1", "col2") AS "subq3"
USING("col2", "col1")
```
```
thread 'test_cross_join_bug2::count_distinct_error' panicked at 'called `Option::unwrap()` on a `None` value', /Users/jonmmease/.cargo/git/checkouts/arrow-datafusion-71ae82d9dec9a01c/556282a/datafusion/core/src/physical_plan/joins/hash_join.rs:923:17
stack backtrace:
...
```

If the order of the columns in the GROUP BY query is also flipped (from `GROUP BY "col1", "col2"` to `GROUP BY "col2", "col1"`) then the query works again:
```sql
SELECT "col1", "col2", coalesce("sum_col3", 0) as sum_col3
FROM (SELECT DISTINCT "col2" FROM "tbl") AS "subq1"
CROSS JOIN (SELECT DISTINCT "col1" FROM "tbl") AS "subq2"
LEFT OUTER JOIN (SELECT "col1", "col2", sum("col3") as "sum_col3" FROM "tbl" GROUP BY "col2", "col1") AS "subq3"
USING("col2", "col1")
```
```
+------+------+----------+
| col1 | col2 | sum_col3 |
+------+------+----------+
| BB   | 5    | 2        |
| BB   | 1    | 0        |
| BB   | 4    | 0        |
| A    | 3    | 2        |
| BB   | 2    | 0        |
| BB   | 6    | 2        |
| BB   | 3    | 0        |
| A    | 2    | 2        |
| A    | 6    | 0        |
| A    | 4    | 2        |
| A    | 1    | 2        |
| A    | 5    | 0        |
+------+------+----------+
```

**To Reproduce**
Here is a Rust test that implements the two queries above:

```rust
#[cfg(test)]
mod test_cross_join_bug2 {
    use std::sync::Arc;
    use datafusion::arrow::array::{ArrayRef, StringArray, UInt64Array};
    use datafusion::arrow::datatypes::{DataType, Field, Schema, SchemaRef};
    use datafusion::arrow::record_batch::RecordBatch;
    use datafusion::arrow::util::pretty::pretty_format_batches;
    use datafusion::datasource::MemTable;
    use datafusion::prelude::SessionContext;

    #[tokio::test]
    async fn count_distinct_error() {
        let col1 = Arc::new(StringArray::from(vec![
            "A", "A", "A", "A",
            "A", "A", "A", "A",
            "BB", "BB", "BB", "BB",
        ])) as ArrayRef;

        let col2 = Arc::new(UInt64Array::from(vec![
            1, 1, 2, 2, 3, 3, 4, 4, 5, 5, 6, 6
        ])) as ArrayRef;

        let col3 = Arc::new(UInt64Array::from(vec![
            1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1
        ])) as ArrayRef;

        let schema = Arc::new(Schema::new(vec![
            Field::new("col1", DataType::Utf8, true),
            Field::new("col2", DataType::UInt64, true),
            Field::new("col3", DataType::UInt64, true),
        ])) as SchemaRef;

        let batch =
            RecordBatch::try_new(schema.clone(), vec![
                col1, col2, col3,
            ]).unwrap();
        let mem_table = MemTable::try_new(schema, vec![vec![batch]]).unwrap();

        // Create context and register table
        let ctx = SessionContext::new();
        ctx.register_table("tbl", Arc::new(mem_table)).unwrap();

        // Pretty print the input table
        let res = ctx
            .sql("SELECT * from tbl")
            .await
            .unwrap()
            .collect()
            .await
            .unwrap();
        let formatted = pretty_format_batches(res.as_slice()).unwrap();
        println!("{}", formatted);

        // Perform query where GROUP BY and USING both specify "col1", "col2" (works)
        let sql1 = r#"
SELECT "col1", "col2", coalesce("sum_col3", 0) as sum_col3
FROM (SELECT DISTINCT "col2" FROM "tbl") AS "subq1"
CROSS JOIN (SELECT DISTINCT "col1" FROM "tbl") AS "subq2"
LEFT OUTER JOIN (SELECT "col1", "col2", sum("col3") as "sum_col3" FROM "tbl" GROUP BY "col1", "col2") AS "subq3"
USING("col1", "col2")
    "#;

        let res = ctx.sql(sql1).await.unwrap().collect().await.unwrap();
        let formatted = pretty_format_batches(res.as_slice()).unwrap();
        println!("{}", formatted);

        // Perform query where GROUP BY specifies "col1", "col2" and USING specifies "col2", "col1" (panic)
        let sql2 = r#"
SELECT "col1", "col2", coalesce("sum_col3", 0) as sum_col3
FROM (SELECT DISTINCT "col2" FROM "tbl") AS "subq1"
CROSS JOIN (SELECT DISTINCT "col1" FROM "tbl") AS "subq2"
LEFT OUTER JOIN (SELECT "col1", "col2", sum("col3") as "sum_col3" FROM "tbl" GROUP BY "col1", "col2") AS "subq3"
USING("col2", "col1")
    "#;

        let res = ctx.sql(sql2).await.unwrap().collect().await.unwrap();
        let formatted = pretty_format_batches(res.as_slice()).unwrap();
        println!("{}", formatted);
    }
}
```

Output
```
+------+------+------+
| col1 | col2 | col3 |
+------+------+------+
| A    | 1    | 1    |
| A    | 1    | 1    |
| A    | 2    | 1    |
| A    | 2    | 1    |
| A    | 3    | 1    |
| A    | 3    | 1    |
| A    | 4    | 1    |
| A    | 4    | 1    |
| BB   | 5    | 1    |
| BB   | 5    | 1    |
| BB   | 6    | 1    |
| BB   | 6    | 1    |
+------+------+------+
+------+------+----------+
| col1 | col2 | sum_col3 |
+------+------+----------+
| A    | 3    | 2        |
| BB   | 2    | 0        |
| BB   | 6    | 2        |
| BB   | 3    | 0        |
| BB   | 5    | 2        |
| BB   | 1    | 0        |
| BB   | 4    | 0        |
| A    | 4    | 2        |
| A    | 2    | 2        |
| A    | 6    | 0        |
| A    | 1    | 2        |
| A    | 5    | 0        |
+------+------+----------+

thread 'test_cross_join_bug2::count_distinct_error' panicked at 'called `Option::unwrap()` on a `None` value', /Users/jonmmease/.cargo/git/checkouts/arrow-datafusion-71ae82d9dec9a01c/556282a/datafusion/core/src/physical_plan/joins/hash_join.rs:923:17
```

**Expected behavior**
These two queries should both complete successfully, regardless of whether the columns are specified in the same order in the GROUP BY and USING constraints.

### PR DESCRIPTION
**Pull request by jonmmease**
# Which issue does this PR close?

Closes #4873.

# Rationale for this change
See #4873 for a failing example and some preliminary investigation.

The issue is in the EnforceDistribution physical optimizer (the query from the issue works without it).  After looking at this for a while, I think I see what's going on.  Normally, the column expressions for a group by have indices that correspond to the schema of the input operation. But for the case of a partial aggregation, these indices need to be updated to correspond to the schema of the partial aggregation instead of the input.

This make sense to me, and it fixes the query, but this is the first time I've looked at the Physical optimizer so someone who's more familiar with that side of things should definitely take a look before merging this.

# Are these changes tested?

I added a test case based on the query reported in the issue.

### CODE REVIEW

File: datafusion/core/src/physical_optimizer/dist_enforcement.rs [hunk 1]
```diff
@@ -431,9 +431,22 @@ fn reorder_aggregate_keys(
...
+                    for (idx, position) in positions.into_iter().enumerate() {
+                        let (group_expr, name) = group_by.expr()[position].clone();
+
+                        // Update the index of columns so that they correspond to the schema of partial_agg
+                        // rather than input_schema.
+                        let group_expr = if let Some(group_col) =
+                            group_expr.as_any().downcast_ref::<Column>()
+                        {
+                            Arc::new(Column::new(group_col.name(), idx))
+                                as Arc<dyn PhysicalExpr>
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/dist_enforcement.rs:445]: I think you can do something like this to get the rust compiler to do the right cast for you:

```suggestion
                                as _
```

However the rest of this module is in the same `as Arc<...>` style so no need to change it in this PR

File: datafusion/core/tests/sql/joins.rs [hunk 2]
```diff
@@ -2810,3 +2810,61 @@ async fn type_coercion_join_with_filter_and_equi_expr() -> Result<()> {
 
     Ok(())
 }
+
+#[tokio::test]
+async fn test_cross_join_to_groupby_with_different_key_ordering() -> Result<()> {
```
Comment (alamb) [inline datafusion/core/tests/sql/joins.rs:2815]: I verified that this test does indeed panic without the test code

```
---- sql::joins::test_cross_join_to_groupby_with_different_key_ordering stdout ----
thread 'sql::joins::test_cross_join_to_groupby_with_different_key_ordering' panicked at 'called `Option::unwrap()` on a `None` value', datafusion/core/src/physical_plan/joins/hash_join.rs:923:17
stack backtrace:
```

**Review by alamb — APPROVED** [review]
Thank you @jonmmease  -- reading the 🕵️  story on https://github.com/apache/arrow-datafusion/issues/4873#issue-1529118115 is fascinating

The fact that this code fixes a bug and has test coverage is quite compelling to me 
 
@mingmwang  and @yahoNanJing , who I think are familiar / authors of this code, can you please take a look if you have time. Also cc @metesynnada

**Comment by mingmwang** [comment]
The Column Index in the physical plan is confusing and error prone, I will take a closer look at this PR.

File: datafusion/core/src/physical_optimizer/dist_enforcement.rs (hunk 1 above)
Comment (mingmwang) [reply datafusion/core/src/physical_optimizer/dist_enforcement.rs:445]: Good catch!!  It is a bug.  

Regarding the fix, can we call this method to get the new `new_group_exprs` ?

`let new_group_exprs: Vec<Arc<dyn PhysicalExpr>> = partial_agg.output_group_expr();`

**Comment by mingmwang** [comment]
Regarding the fix, I think you can call below methods to simply the logic. I had test it on my local and it works.

```rust
                    // Build new group expressions that correspond to the output of partial_agg
                    let new_final_group: Vec<Arc<dyn PhysicalExpr>> =
                        partial_agg.output_group_expr();
                    let new_group_by= PhysicalGroupBy::new_single(
                        new_final_group
                            .iter()
                            .enumerate()
                            .map(|(i, expr)| (expr.clone(), partial_agg.group_expr().expr()[i].1.clone()))
                            .collect(),
                    );
```

**Review by metesynnada — COMMENTED** [review]
LGTM

**Comment by jonmmease** [comment]
Thanks for taking a look @alamb @mingmwang @metesynnada!  I made the simplification @mingmwang suggested in [f32a8d9](https://github.com/apache/arrow-datafusion/pull/4878/commits/f32a8d922cc300ebafb8b1eab000a3fd3b82632c).

**Comment by metesynnada** [comment]
LGTM, good work 😀

**Comment by alamb** [comment]
Thank you everyone for your help!

**Pull request merged**